src/GafferUI/GraphGadget.cpp

```diff
@@ -223,6 +232,230 @@ class SetPositionsAction : public Gaffer::Action
+					activeWalk( input, contextProcessorNode->inPlug(), newContext.get(), canceller, activePlugs, activeNodes, plugContextsVisited );
+				}
+				handledBySpecialCase.push_back( contextProcessorNode->inPlug() );
+			}
+		}
+
+		if( const Gaffer::Loop *loopNode = runTimeCast<const Gaffer::Loop>( node ) )
+		{
+			if(
+				!firstVisit &&
+				( connectionStart == loopNode->previousPlug() ||  connectionStart->parent() == loopNode->previousPlug() )
+			)
+			{
+				// In any normal circumstance, if we've already visited this loop, that means we will
+				// have already done all needed traversal, and just stop here.
+				//
+				// Stopping here avoids leaking out the traversal from inside the loop where we don't
+				// know the context.
+				//
+				// There are some weird cases where this might not be totally accurate - it's technically
+				// possible to make a graph that uses the inner nodes of a loop both through the loop, and
+				// also through manually using a ContextVariables node to set loop:index - this is a very
+				// rare special case though
+				return;
+			}
+
+			// The input plug of the loop is a normal input, and will be handled below
+
+			// The next plug of the loop is evaluated in many different contexts - instead of evaluating all
+			// of them, pass a null context which will trigger a conservative evaluation of all possible inputs
+			const Gaffer::Plug *input2 = loopNode->nextPlug()->getInput();
+			if( input2 )
+			{
+				activeWalk( input2, loopNode->nextPlug(), nullptr, canceller, activePlugs, activeNodes, plugContextsVisited );
+			}
+			handledBySpecialCase.push_back( loopNode->nextPlug() );
+		}
+	}
+	catch( const Gaffer::ProcessException &e )
+	{
+		// If there's an error in the graph, we can't figure out specifically which nodes contribute,
+		// fall back to showing all possible inputs as active
+		context = nullptr;
+		handledBySpecialCase.clear();
+
+		IECore::msg( IECore::Msg::Warning, "Gaffer", std::string( "Error during graph active state visualisation: " ) + e.what() );
+	}
+
+	// For default nodes
+	for( auto &inSourcePlug : Gaffer::Plug::RecursiveInputRange( *node ) )
+	{
+		bool handled = false;
+		for( const Gaffer::Plug* handledPlug : handledBySpecialCase )
+		{
+			// TODO - using isAncestor here is pretty ugly, alternative would be doing recursive
+			// traversal ourselves so we could terminate if we encounter a plug that is handled?
+			if( handledPlug == inSourcePlug || handledPlug->isAncestorOf( inSourcePlug.get() ) )
+			{
+				handled = true;
+				break;
+			}
+		}
+
+		if( handled )
+		{
+			continue;
+		}
+
+		Gaffer::Plug *input = inSourcePlug->getInput();
+		Gaffer::Plug *inSourceParentPlug = runTimeCast<Gaffer::Plug>( inSourcePlug->parent() );
+		if( input )
+		{
+			if( inSourceParentPlug && input->parent() == inSourceParentPlug->getInput() )
+			{
+				// Parents are connected, don't need to trace individual children
+				continue;
+			}
```

Comment:
Addresed in ae63d29.  I did note while working on this that current tests are just on FloatPlug rather than a hierarchical plug like ScenePlug, so this code is not tested well by the unit tests.  Maybe need to add some more.